### PR TITLE
Perf/wgpu/autotune

### PR DESCRIPTION
**Pull request by louisfd**
Added Autotune for matmul

### CODE REVIEW

**Comment by louisfd** [comment]
I deactivated the default on the autotune compile because it makes the wgpu test suite extremely slow. 
We should inspect the cause of the slowness before putting it to default. I'll open an issue. 
Nevertheless, I think this PR can be merged as it does not slow down the test suite if the flag is not set.

File: burn-wgpu/src/benchmark.rs [hunk 1]
```diff
@@ -118,11 +130,11 @@ macro_rules! run_benchmark {
...
-            println!(
-                "OpenGL - {}{}",
-                Benchmark::<burn_wgpu::OpenGl>::name(&$bench),
-                Benchmark::<burn_wgpu::OpenGl>::run(&$bench, &WgpuDevice::DiscreteGpu(0))
-            );
+            // println!(
+            //     "OpenGL - {}{}",
+            //     Benchmark::<burn_wgpu::OpenGl>::name(&$bench),
+            //     Benchmark::<burn_wgpu::OpenGl>::run(&$bench, &WgpuDevice::DiscreteGpu(0))
+            // );
```
Comment (nathanielsimard) [inline burn-wgpu/src/benchmark.rs:137]: We should just delete this.

File: burn-wgpu/src/benchmark.rs [hunk 2]
```diff
@@ -71,8 +83,8 @@ pub trait Benchmark<G: GraphicsApi> {
     fn name(&self) -> String;
     /// Run the benchmark a number of times.
     fn run(&self, device: &WgpuDevice) -> BenchmarkResult
-    where
-        Self: Sized,
+// where
+        // Self: Sized,
```
Comment (nathanielsimard) [inline burn-wgpu/src/benchmark.rs:87]: Dead code

File: burn-wgpu/src/context/base.rs [hunk 3]
```diff
@@ -168,7 +171,7 @@ impl Context {
     /// Compile a kernel template if not present in the cache.
     pub fn compile_static<K: StaticKernel>(&self) -> Arc<ComputePipeline> {
         let mut cache = self.cache.lock();
-        let template_id = Key::Static(TypeId::of::<K>());
+        let template_id = ComputeKey::Static(TypeId::of::<K>());
```
Comment (nathanielsimard) [inline burn-wgpu/src/context/base.rs:174]: Let's rename `ComputeKey` to `TemplateKey`

File: burn-wgpu/src/kernel/matmul/tune.rs [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+pub fn tune<G: GraphicsApi, E, const D: usize>(
+    lhs: WgpuTensor<E, D>,
+    rhs: WgpuTensor<E, D>,
+) -> WgpuTensor<E, D>
+where
+    E: WgpuElement + FloatElement,
+{
+    let (shape_lhs, shape_rhs) = calculate_benchmark_shapes(lhs.shape.clone(), rhs.shape.clone());
+    let id = AutoTuneKey::new(
+        vec![Vec::from(shape_lhs.dims), Vec::from(shape_rhs.dims)],
```
Comment (nathanielsimard) [inline burn-wgpu/src/kernel/matmul/tune.rs:165]: I think there is a problem, we will recompute the benchmarks for tensors of different rank. we should only add M, N and K in the list of usize.

File: burn-wgpu/src/kernel/matmul/tune.rs [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+    E: WgpuElement + FloatElement,
+{
+    let matmul_benchmark =
+        |func: AutoTuneFunction<(WgpuTensor<E, D>, WgpuTensor<E, D>), WgpuTensor<E, D>>| {
+            Tunable::<G, _, _>::new(
+                func.clone(),
+                Arc::new(MatmulBenchmark::new(
+                    shape_lhs.clone(),
+                    shape_rhs.clone(),
+                    10,
```
Comment (nathanielsimard) [inline burn-wgpu/src/kernel/matmul/tune.rs:222]: Should be 5 I think.

File: burn-wgpu/src/ops/float_ops.rs [hunk 6]
```diff
@@ -139,6 +139,10 @@ where
         lhs: FloatTensor<Self, D>,
         rhs: FloatTensor<Self, D>,
     ) -> FloatTensor<Self, D> {
+        #[cfg(feature = "autotune")]
+        return kernel::matmul::tune::<G, F, D>(lhs, rhs);
```
Comment (nathanielsimard) [inline burn-wgpu/src/ops/float_ops.rs:143]: I think we should keep the autotune default, but only run if not in test mode.

`#[cfg(all(feature = "autotune", not(test))]`

**Review by nathanielsimard — CHANGES_REQUESTED** [review]

File: burn-wgpu/src/kernel/matmul/tune.rs [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+        return kernel::matmul::matmul_tiling_2d_default(lhs, rhs);
+    }
+
+    let (shape_lhs, shape_rhs) = calculate_benchmark_shapes(lhs.shape.clone(), rhs.shape.clone());
+    let id = AutoTuneKey::new(
+        vec![
+            shape_lhs.dims[D - 2..].to_vec(),
+            shape_rhs.dims[D - 2..].to_vec(),
+        ],
+        format!("matmul {}", E::type_name()),
```
Comment (nathanielsimard) [inline burn-wgpu/src/kernel/matmul/tune.rs:253]: Mettre device dans id.

**Review by nathanielsimard — APPROVED** [review]

**Pull request merged**